### PR TITLE
[Dev] MuonClip support (non-split version) on dev branch

### PR DESCRIPTION
**Pull request by BoxiangW**
# What does this PR do ?
Add MuonClip support, sister PR to main branch is https://github.com/NVIDIA/Megatron-LM/pull/1929.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/arguments.py [hunk 1]
```diff
@@ -1003,6 +1003,14 @@ def validate_args(args, defaults={}):
     if args.add_bias_linear:
         args.add_qkv_bias = True
 
+    if args.qk_clip:
+        assert is_te_min_version("2.9.0"), \
+            '--qk-clip is only supported with TE >= 2.9.0.'
+        assert 0.0 < args.qk_clip_alpha < 1.0, \
+            '--qk-clip-balancing-alpha must be between 0.0 and 1.0 when using --qk-clip.'
```
Comment (FDecaYed) [inline megatron/training/arguments.py]: qk-clip-balancing-alpha and qk-clip-balancing-threshold in assert message is not same as option name

File: megatron/core/transformer/attention.py [hunk 2]
```diff
@@ -1227,6 +1234,104 @@ def set_for_recompute_input_layernorm(self):
...
+                but {grouped_max_attn_logits.shape}"
+            qk_clip_balancing_eta = torch.clamp(
+                self.config.qk_clip_threshold / grouped_max_attn_logits, max=1.0
+            ).view(self.num_query_groups_per_partition, 1, 1)
+            assert torch.all(qk_clip_balancing_eta <= 1.0)
+
+            # Handle different weight access patterns (main_param vs direct access)
+            if hasattr(self.linear_qkv.weight, 'main_param'):
+                weight = self.linear_qkv.weight.main_param.data
+                weight_shape = weight.shape
```
Comment (FDecaYed) [inline megatron/core/transformer/attention.py]: seems not used

File: megatron/core/transformer/attention.py [hunk 3]
```diff
@@ -1227,6 +1234,104 @@ def set_for_recompute_input_layernorm(self):
...
+            )
+            weight_k = weight_k * torch.pow(qk_clip_balancing_eta, 1 - self.config.qk_clip_alpha)
+
+            # Concatenate back and reshape to original shape
+            weight_updated = torch.cat([weight_q, weight_k, weight_v], dim=1)
+            weight_updated = weight_updated.view(
+                self.query_projection_size + 2 * self.kv_projection_size, -1
+            )
+
+            # Apply the updated weights
```
Comment (FDecaYed) [inline megatron/core/transformer/attention.py]: why not just do inplace mul_()?

File: megatron/training/training.py [hunk 4]
```diff
@@ -1404,6 +1405,13 @@ def train_step(forward_step_func, data_iterator, model, optimizer, opt_param_sch
 
     timers('optimizer', log_level=1).start(barrier=args.barrier_with_L1_time)
     update_successful, grad_norm, num_zeros_in_grad = optimizer.step()
+
+    # get max attention logit for logging and run clip_qk()
+    # Part of MuonClip Optimizer step
+    log_max_attention_logit = 0
```
Comment (FDecaYed) [inline megatron/training/training.py:1412]: I'd say we make logging decoupled with clipping and allow turning on/off. This way user can train their model, knowing logits and use it to determine clipping later on
can also open a separate PR to track this and implement later

**Review by FDecaYed — APPROVED** [review]

**Review by FDecaYed — CHANGES_REQUESTED** [review]
Ok. successfully removed it from merge queue. 
Thinking about this again, it seems just clipping main_param and not copying to model param is wrong?

**Comment by FDecaYed** [comment]
issues:
- logits max need to be synced between DP ranks be fore clipping
- for ranks have main_param, both main_param and param need to be updated
- above 2 changes combined will also properly support layerwise wrapper(without needing another sync)

File: megatron/core/transformer/attention.py (hunk 2 above)
Comment (BoxiangW) [reply megatron/core/transformer/attention.py]: Fixed, thanks for catching it. Not sure why lint didn't find it

File: megatron/core/optimizer/qk_clip.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    with torch.no_grad():
+        log_max_attention_logit = 0
+        for model_chunk in model:
+            for transformer_layer in model_chunk.module.module.decoder.layers:
+                if hasattr(transformer_layer.self_attention, 'clip_qk'):
+                    torch.distributed.all_reduce(
+                        transformer_layer.self_attention.core_attention.current_max_attn_logits,
+                        op=torch.distributed.ReduceOp.MAX,
+                        group=mpu.get_data_parallel_group(),
```
Comment (FDecaYed) [inline megatron/core/optimizer/qk_clip.py]: I feel this should be dp_cp group?

File: megatron/core/transformer/attention.py [hunk 6]
```diff
@@ -1239,6 +1246,103 @@ def set_for_recompute_input_layernorm(self):
...
+                but {grouped_max_attn_logits.shape}"
+            self.qk_clip_balancing_eta = torch.clamp(
+                self.config.qk_clip_threshold / grouped_max_attn_logits, max=1.0
+            ).view(self.num_query_groups_per_partition, 1, 1)
+            assert torch.all(self.qk_clip_balancing_eta <= 1.0)
+
+            # Handle different weight access patterns (main_param vs direct access)
+            if hasattr(self.linear_qkv.weight, 'main_param'):
+                self.linear_qkv.weight.main_param.data.copy_(
+                    self._clip_linear_qkv(self.linear_qkv.weight.main_param.data)
```
Comment (FDecaYed) [inline megatron/core/transformer/attention.py:1290]: probably ok this way but put my thought here:

clipping both main/model weight might result in slight higher mismatch than just lower precision rounding, not sure about exact effect on training.

the 'correct' thing should be clip main param and copy to model param. but if we're not doing another communication, then some rank might not have access to main param. if only rank with main param copy, then mismatch model param between ranks feels even worse

another choice is clip model and copy to main, which feels also worse than current approach

File: megatron/training/arguments.py [hunk 7]
```diff
@@ -3286,7 +3310,7 @@ def _add_mla_args(parser):
     group = parser.add_argument_group(title="mla")
     group.add_argument('--q-lora-rank', type=int, default=None,
                        help="Rank of Query tensor's low rank representation.")
-    group.add_argument('--kv-lora-rank', type=int, default=32,
+    group.add_argument('--kv-lora-rank', type=int, default=512,
```
Comment (FDecaYed) [inline megatron/training/arguments.py]: any reason to mix this change in?

**Review by FDecaYed — CHANGES_REQUESTED** [review]
Requesting small changes, see comments. And I'm assuming clipping/weight shaping detail is correct without testing.
Other than that, code logic looks good to me.

File: megatron/core/optimizer/qk_clip.py (hunk 5 above)
Comment (BoxiangW) [reply megatron/core/optimizer/qk_clip.py]: Yes, fixed, thanks for catching it

File: megatron/core/transformer/attention.py (hunk 6 above)
Comment (BoxiangW) [reply megatron/core/transformer/attention.py:1290]: Maybe I can create another PR for moving it inside `optimizer.step()` so that I can change only on main_param before the FP32->bf16 cast.

File: megatron/training/arguments.py (hunk 7 above)
Comment (BoxiangW) [reply megatron/training/arguments.py]: Not related to this PR, but this 32 default value is kind of small in general. Reverted in this PR

File: megatron/core/transformer/attention.py (hunk 6 above)
Comment (FDecaYed) [reply megatron/core/transformer/attention.py:1290]: Let's run some benchmark and profiling after this. In my opinion we'll have to make a version later that mixes things inside optimizer.step() as perf optimization.
